packages/test/test-end-to-end-tests/src/test/layerCompat.spec.ts

```diff
@@ -0,0 +1,179 @@
+} from "@fluidframework/container-loader/internal";
+import {
+	FluidErrorTypes,
+	type ITelemetryBaseProperties,
+} from "@fluidframework/core-interfaces/internal";
+import { localDriverCompatDetailsForLoader } from "@fluidframework/local-driver/internal";
+import { UsageError } from "@fluidframework/telemetry-utils/internal";
+import { ITestObjectProvider } from "@fluidframework/test-utils/internal";
+
+type ILayerCompatSupportRequirementsOverride = Omit<
+	ILayerCompatSupportRequirements,
+	"requiredFeatures" | "minSupportedGeneration"
+> & {
+	requiredFeatures: (typeof driverSupportRequirements)["requiredFeatures"];
+	minSupportedGeneration: (typeof driverSupportRequirements)["minSupportedGeneration"];
+};
+
+type LayerType = "Loader" | "Driver";
+
+function validateFailureProperties(
+	error: Error,
+	isGenerationCompatible: boolean,
+	layerTypes: LayerType[],
+	minSupportedGeneration: number,
+	unsupportedFeatures?: string[],
+): boolean {
+	assert(error instanceof UsageError, "The error should be a UsageError");
+	assert.strictEqual(
+		error.errorType,
+		FluidErrorTypes.usageError,
+		"Error type should be usageError",
+	);
+	const telemetryProps = error.getTelemetryProperties();
+	assert(typeof telemetryProps.errorDetails === "string", "Error details should be present");
+	const properties = JSON.parse(telemetryProps.errorDetails) as ITelemetryBaseProperties;
+	assert.strictEqual(
+		properties.isGenerationCompatible,
+		isGenerationCompatible,
+		"Generation compatibility not as expected",
+	);
+	assert.strictEqual(
+		properties.minSupportedGeneration,
+		minSupportedGeneration,
+		"Minimum supported generation not as expected",
+	);
+	assert.deepStrictEqual(
+		properties.unsupportedFeatures,
+		unsupportedFeatures,
+		"Unsupported features not as expected",
+	);
+
+	for (const layerType of layerTypes) {
+		switch (layerType) {
+			case "Loader":
+				assert.strictEqual(
+					properties.loaderVersion,
+					loaderCoreCompatDetails.pkgVersion,
+					"Runtime version not as expected",
+				);
+				assert.strictEqual(
+					properties.loaderGeneration,
+					loaderCoreCompatDetails.generation,
+					"Runtime generation not as expected",
+				);
+				break;
+			case "Driver":
+				assert.strictEqual(
+					properties.driverVersion,
+					localDriverCompatDetailsForLoader.pkgVersion,
+					"Driver version not as expected",
+				);
+				assert.strictEqual(
+					properties.driverGeneration,
+					localDriverCompatDetailsForLoader.generation,
+					"Driver generation not as expected",
+				);
+				break;
+			default:
+				assert.fail(`Unexpected layer type: ${layerType}`);
+		}
+	}
+	return true;
+}
+
+describeCompat("Layer compatibility", "NoCompat", (getTestObjectProvider) => {
+	let provider: ITestObjectProvider;
+	beforeEach("getTestObjectProvider", function () {
+		provider = getTestObjectProvider();
+
+		if (provider.driver.type !== "local") {
+			this.skip();
+		}
+	});
+
+	describe("Loader / Driver compat", () => {
```

Comment:
Want to move Loader / Runtime compat into this same test file?  Maybe in a separate PR.

Reply:
Yeah, I am planning to eventually add end-to-end tests for other layers here too.